Test case test_worker_node_restart_during_pvc_expansion failure

Test case test_worker_node_restart_during_pvc_expansion is failing with the below error while waiting for volume expansion to complete. The capacity of the RBD PVC is not increased within the given wait time of 240 seconds but pending on filesystem resize. This happens if the restarted node is not in Ready state within 240 seconds. Also the app pods on the restarted pod will get terminated and will be re-created on other nodes.


```
        # Verify pvc expansion status
        for pvc_obj in self.pvcs:
>           assert (
                pvc_obj.expand_proc.result()
            ), f"Expansion failed for PVC {pvc_obj.name}"

tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py:95: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.6/concurrent/futures/_base.py:425: in result
    return self.__get_result()
/usr/lib64/python3.6/concurrent/futures/_base.py:384: in __get_result
    raise self._exception
/usr/lib64/python3.6/concurrent/futures/thread.py:56: in run
    result = self.fn(*self.args, **self.kwargs)
ocs_ci/ocs/resources/pvc.py:154: in resize_pvc
    for pvc_data in TimeoutSampler(240, 2, self.get):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ocs_ci.utility.utils.TimeoutSampler object at 0x7f40c03b6d68>

    def __iter__(self):
        if self.start_time is None:
            self.start_time = time.time()
        while True:
            self.last_sample_time = time.time()
            try:
                yield self.func(*self.func_args, **self.func_kwargs)
            except Exception as ex:
                msg = f"Exception raised during iteration: {ex}"
                logging.error(msg)
            if self.timeout < (time.time() - self.start_time):
>               raise self.timeout_exc_cls(*self.timeout_exc_args)
E               ocs_ci.ocs.exceptions.TimeoutExpiredError: Timed Out: (240,)

ocs_ci/utility/utils.py:828: TimeoutExpiredError
```

--- a/ocs_ci/ocs/platform_nodes.py
+++ b/ocs_ci/ocs/platform_nodes.py
@@ -1742,7 +1742,7 @@ class AZURENodes(NodesBase):
                 node_names=node_names, status=constants.NODE_READY, timeout=timeout
             )
 
-    def restart_nodes(self, nodes, timeout=540, wait=True):
+    def restart_nodes(self, nodes, timeout=900, wait=True):
         """
         Restart Azure vm instances
 

--- a/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
+++ b/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
@@ -43,8 +43,8 @@ class TestNodeRestartDuringPvcExpansion(ManageTest):
         self.pvcs, self.pods = create_pvcs_and_pods(
             pvc_size=4,
             pods_for_rwx=2,
-            num_of_rbd_pvc=15,
-            num_of_cephfs_pvc=10,
+            num_of_rbd_pvc=9,
+            num_of_cephfs_pvc=6,
             deployment_config=True,
         )
 

--- a/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
+++ b/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
@@ -137,6 +137,7 @@ class TestNodeRestartDuringPvcExpansion(ManageTest):
                 size="6G",
                 runtime=30,
                 fio_filename=f"{pod_obj.name}_file",
+                end_fsync=1,
             )
 
         log.info("Wait for IO to complete on all pods")

--- a/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
+++ b/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
@@ -99,7 +99,7 @@ class TestNodeRestartDuringPvcExpansion(ManageTest):
         new_pods_list = []
         for pod_obj in self.pods:
             new_pods = get_all_pods(
-                namespace=self.namespace,
+                namespace=pod_obj.namespace,
                 selector=[pod_obj.labels.get("deploymentconfig")],
                 selector_label="deploymentconfig",
                 wait=True,

--- a/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
+++ b/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
@@ -3,6 +3,7 @@ import pytest
 from concurrent.futures import ThreadPoolExecutor
 
 from ocs_ci.ocs import constants, node
+from ocs_ci.ocs.resources.pod import get_all_pods
 from ocs_ci.utility.utils import ceph_health_check
 from ocs_ci.helpers.helpers import wait_for_resource_state
 from ocs_ci.framework.testlib import (
@@ -40,7 +41,11 @@ class TestNodeRestartDuringPvcExpansion(ManageTest):
 
         """
         self.pvcs, self.pods = create_pvcs_and_pods(
-            pvc_size=4, pods_for_rwx=2, num_of_rbd_pvc=15, num_of_cephfs_pvc=10
+            pvc_size=4,
+            pods_for_rwx=2,
+            num_of_rbd_pvc=15,
+            num_of_cephfs_pvc=10,
+            deployment_config=True,
         )
 
     @pytest.fixture(autouse=True)
@@ -97,9 +102,25 @@ class TestNodeRestartDuringPvcExpansion(ManageTest):
             ), f"Expansion failed for PVC {pvc_obj.name}"
         log.info("PVC expansion was successful on all PVCs")
 
+        # Find respun pods
+        new_pods_list = []
+        for pod_obj in self.pods:
+            new_pods = get_all_pods(
+                namespace=self.namespace,
+                selector=[pod_obj.labels.get("deploymentconfig")],
+                selector_label="deploymentconfig",
+                wait=True,
+            )
+            for pod_ob in new_pods:
+                pod_ob.pvc = pod_obj.pvc
+            new_pods_list.extend(new_pods)
+        assert len(new_pods_list) == len(
+            self.pods
+        ), "Couldn't find all pods after node reboot"
+
         # Run IO
         log.info("Run IO after PVC expansion.")
-        for pod_obj in self.pods:
+        for pod_obj in new_pods_list:
             wait_for_resource_state(pod_obj, constants.STATUS_RUNNING)
             storage_type = "block" if pod_obj.pvc.volume_mode == "Block" else "fs"
             pod_obj.io_proc = executor.submit(
@@ -111,7 +132,7 @@ class TestNodeRestartDuringPvcExpansion(ManageTest):
             )
 
         log.info("Wait for IO to complete on all pods")
-        for pod_obj in self.pods:
+        for pod_obj in new_pods_list:
             pod_obj.io_proc.result()
             fio_result = pod_obj.get_fio_results()
             err_count = fio_result.get("jobs")[0].get("error")

--- a/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
+++ b/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
@@ -96,16 +96,20 @@ class TestNodeRestartDuringPvcExpansion(ManageTest):
 
         # Find respun pods
         new_pods_list = []
+        wait_to_stabilize = True
         for pod_obj in self.pods:
             new_pods = get_all_pods(
                 namespace=pod_obj.namespace,
                 selector=[pod_obj.labels.get("deploymentconfig")],
                 selector_label="deploymentconfig",
-                wait=False,
+                wait=wait_to_stabilize,
             )
             for pod_ob in new_pods:
                 pod_ob.pvc = pod_obj.pvc
             new_pods_list.extend(new_pods)
+            # Given enough time for pods to respin. So wait time
+            # is not needed for further iterations
+            wait_to_stabilize = False
         assert len(new_pods_list) == len(
             self.pods
         ), "Couldn't find all pods after node reboot"

--- a/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
+++ b/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
@@ -122,7 +122,7 @@ class TestNodeRestartDuringPvcExpansion(ManageTest):
             capacity = pvc_obj.get().get("status").get("capacity").get("storage")
             assert capacity == f"{pvc_size_expanded}Gi", (
                 f"Capacity of PVC {pvc_obj.name} is not {pvc_size_expanded}Gi as "
-                f"expected, but {capacity}. Retrying."
+                f"expected, but {capacity}."
             )
         log.info("PVC expansion was successful on all PVCs")
 

--- a/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
+++ b/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
@@ -82,7 +82,7 @@ class TestNodeRestartDuringPvcExpansion(ManageTest):
         for pvc_obj in self.pvcs:
             log.info(f"Expanding size of PVC {pvc_obj.name} to {pvc_size_expanded}G")
             pvc_obj.expand_proc = executor.submit(
-                pvc_obj.resize_pvc, pvc_size_expanded, True
+                pvc_obj.resize_pvc, pvc_size_expanded, False
             )
 
         # Check result of node 'restart_nodes'
@@ -94,13 +94,6 @@ class TestNodeRestartDuringPvcExpansion(ManageTest):
             status=constants.NODE_READY,
             timeout=300,
         )
-
-        # Verify pvc expansion status
-        for pvc_obj in self.pvcs:
-            assert (
-                pvc_obj.expand_proc.result()
-            ), f"Expansion failed for PVC {pvc_obj.name}"
-        log.info("PVC expansion was successful on all PVCs")
 
         # Find respun pods
         new_pods_list = []
@@ -117,6 +110,18 @@ class TestNodeRestartDuringPvcExpansion(ManageTest):
         assert len(new_pods_list) == len(
             self.pods
         ), "Couldn't find all pods after node reboot"
+
+        # Verify PVC expansion status
+        for pvc_obj in self.pvcs:
+            assert (
+                pvc_obj.expand_proc.result()
+            ), f"Expansion failed for PVC {pvc_obj.name}"
+            capacity = pvc_obj.get().get("status").get("capacity").get("storage")
+            assert capacity == f"{pvc_size_expanded}Gi", (
+                f"Capacity of PVC {pvc_obj.name} is not {pvc_size_expanded}Gi as "
+                f"expected, but {capacity}. Retrying."
+            )
+        log.info("PVC expansion was successful on all PVCs")
 
         # Run IO
         log.info("Run IO after PVC expansion.")

--- a/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
+++ b/tests/manage/pv_services/pvc_resize/test_node_restart_during_pvc_expansion.py
@@ -42,9 +42,8 @@ class TestNodeRestartDuringPvcExpansion(ManageTest):
         """
         self.pvcs, self.pods = create_pvcs_and_pods(
             pvc_size=4,
-            pods_for_rwx=2,
-            num_of_rbd_pvc=9,
-            num_of_cephfs_pvc=6,
+            num_of_rbd_pvc=12,
+            num_of_cephfs_pvc=8,
             deployment_config=True,
         )
 
@@ -102,7 +101,7 @@ class TestNodeRestartDuringPvcExpansion(ManageTest):
                 namespace=pod_obj.namespace,
                 selector=[pod_obj.labels.get("deploymentconfig")],
                 selector_label="deploymentconfig",
-                wait=True,
+                wait=False,
             )
             for pod_ob in new_pods:
                 pod_ob.pvc = pod_obj.pvc

